src/libraries/System.Linq/src/System/Linq/SequenceEqual.cs

```diff
@@ -29,42 +30,57 @@ public static bool SequenceEqual<TSource>(this IEnumerable<TSource> first, IEnum
                     return ((ReadOnlySpan<TSource>)firstArray).SequenceEqual(secondArray, comparer);
                 }
 
+                if (first is List<TSource> firstList && second is List<TSource> secondList)
+                {
+                    return CollectionsMarshal.AsSpan(firstList).SequenceEqual(CollectionsMarshal.AsSpan(secondList), comparer);
+                }
+
                 if (firstCol.Count != secondCol.Count)
                 {
                     return false;
                 }
 
-                if (firstCol is IList<TSource> firstList && secondCol is IList<TSource> secondList)
+                if (firstCol is IList<TSource> firstIList && secondCol is IList<TSource> secondIList)
```

Comment:
What's the benefit of the rest of the changes?

Reply:
Do you mean the naming change or the moving the loops to local methods?

In case of the former - there was a variable name conflict (reverted now).

In case of the latter, it is to allow the JIT to inline the method and optimize away the checks when .SequenceEqual is called on the exact type. I'm hoping to further evaluate codegen once https://github.com/dotnet/runtime/pull/97005 is merged to see what the exact form is necessary for `list1.SequenceEqual(list2)` so that it is considered to be profitable to get inlined.

Reply:
I suspect optimizing for inlining SequenceEqual will have an overall negative impact, once code size is factored in, and many, many uses are not on a concrete type visible to the JIT even with inlining.

Reply:
Which is why it needs to be written in a way where the type checks are seen as foldable which raises the inlining profitability and ends up in the SequenceEqual being inlined for concrete types (making the call free) and not for some intermediate cases or cases where it is not known that tests against `ICollection<T>` will fail (but still with a marginal cost of a single branch for `ICollection<T>` types).